### PR TITLE
[DEV-1756] Implement drug type for Complex and EntitySet in CuratorTool

### PR DESCRIPTION
**Pull request by lbeckman314**
Tracking [[DEV-1756] Implement drug type for Complex and EntitySet in CuratorTool](https://reactome.atlassian.net/browse/DEV-1756).

There is likely room for more unit and integration tests (like testing complexes, nested drug members, etc.). Will update!

Let me know any areas for improvement!

Screen Shot 2019-10-16 at 5 21 07 PM
Screen Shot 2019-10-16 at 5 21 21 PM

### CODE REVIEW

**Comment by lbeckman314** [comment]
TODO
- [ ] Drug color/labeling for EntitySets erases double line border and rounded borders.

File: src/org/gk/elv/ElvReactionEditHelper.java [hunk 1]
```diff
@@ -239,6 +239,7 @@ private Renderable getRenderableForPE(GKInstance pe,
             rPe = zoomableEditor.getFreeFormObject(pe);
         if (rPe == null)
             rPe = zoomableEditor.insertInstance(pe);
+        System.out.println("rPe: " + rPe);
```
Comment (guanmingwu) [inline src/org/gk/elv/ElvReactionEditHelper.java]: Please delete this line in a production setting.

File: src/org/gk/elv/InstanceZoomablePathwayEditor.java [hunk 2]
```diff
@@ -191,7 +188,8 @@ public boolean isUsedAsDrawingTool() {
     private void init() {
         // Link this selection to other displays
         pathwayEditor.getSelectionModel().addGraphEditorActionListener(new GraphEditorActionListener() {
-            public void graphEditorAction(GraphEditorActionEvent e) {
+            @Override
+			public void graphEditorAction(GraphEditorActionEvent e) {
```
Comment (guanmingwu) [inline src/org/gk/elv/InstanceZoomablePathwayEditor.java]: Hi Liam, Please configure your Eclipse so that your tab is automatically converted into 4 spaces. Otherwise, we see something like this: @Overide and public is not in the same column. By the way, why did you bother to add @Override here? There is no need to spend your time on this.

File: src/org/gk/elv/InstanceZoomablePathwayEditor.java [hunk 3]
```diff
@@ -553,10 +552,11 @@ public Renderable insertInstance(GKInstance instance) {
             Renderable container = pathwayEditor.getRenderable();
             // Don't set container so that the following simpleConvert()
             // will not add the converted instance directly to this container.
-            // Let the insert statments figure this out.
+            // Let the insert statements figure this out.
             //handler.setContainer(container);
             handler.setContainer(null);
             Renderable r = handler.simpleConvert(instance);
+//            Renderable r = InstanceToRenderableConverter.convertToNode(instance, false);
```
Comment (guanmingwu) [inline src/org/gk/elv/InstanceZoomablePathwayEditor.java]: Please remember to remove your test code before push!

File: src/org/gk/elv/InstanceZoomablePathwayEditor.java [hunk 4]
```diff
@@ -63,17 +63,14 @@
...
 import org.gk.render.RenderUtility;
 import org.gk.render.Renderable;
 import org.gk.render.RenderableComplex;
-import org.gk.render.RenderableFeature;
 import org.gk.render.RenderablePathway;
 import org.gk.render.RenderableReaction;
 import org.gk.render.RenderableRegistry;
-import org.gk.schema.GKSchemaClass;
 import org.gk.schema.SchemaClass;
 import org.gk.util.DialogControlPane;
```
Comment (guanmingwu) [inline src/org/gk/elv/InstanceZoomablePathwayEditor.java]: This is a class you should not touch actually. Please focus on what you need to do.

File: src/org/gk/elv/SimpleInstanceTransferHandler.java [hunk 5]
```diff
@@ -39,7 +39,8 @@
     public SimpleInstanceTransferHandler() {
```
Comment (guanmingwu) [inline src/org/gk/elv/SimpleInstanceTransferHandler.java]: I don't think this class is related to what you are doing.

File: src/org/gk/gkCurator/authorTool/ComplexInstanceHandler.java [hunk 6]
```diff
@@ -10,20 +10,29 @@
...
 import org.gk.model.GKInstance;
 import org.gk.model.ReactomeJavaConstants;
+import org.gk.property.SearchDBTypeHelper;
 import org.gk.render.InstanceToRenderableConverter;
 import org.gk.render.Renderable;
 import org.gk.render.RenderableComplex;
 import org.gk.render.RenderableFactory;
 
 public class ComplexInstanceHandler extends InstanceHandler {
+    private SearchDBTypeHelper typeHelper;
```
Comment (guanmingwu) [inline src/org/gk/gkCurator/authorTool/ComplexInstanceHandler.java]: By adding this attribute, you are begging to create a new sup-class to which ComplexInstanceHandler and EntityInstanceHandler can subclass to. Please note that code duplications.

File: src/org/gk/model/InstanceUtilities.java [hunk 7]
```diff
@@ -1394,4 +1402,92 @@ else if (instance.getSchemClass().isa(ReactomeJavaConstants.Pathway))
...
+    	 if (instance == null)
+    		 return false;
+
+    	 SchemaClass schemaClass = instance.getSchemClass();
+    	 // Check if instance is a drug.
+    	 if (schemaClass.isa(ReactomeJavaConstants.Drug))
+    		 return true;
+
+    	 List<String> validClasses = Arrays.asList(ReactomeJavaConstants.EntitySet,
+												   ReactomeJavaConstants.Complex);
```
Comment (guanmingwu) [inline src/org/gk/model/InstanceUtilities.java]: Please adjust your Eclipse setting so that tab is converted into spaces.

File: src/org/gk/model/InstanceUtilities.java [hunk 8]
```diff
@@ -1394,4 +1402,92 @@ else if (instance.getSchemClass().isa(ReactomeJavaConstants.Pathway))
...
+    	 // Check id instance is an EntitySet or Complex.
+    	 if (validClasses.stream().noneMatch(schemaClass::isa))
+			 return false;
+    	 
+    	 // If EntitySet has a "hasMember" or "hasCandidate" attribute then recursively iterate over them.
+    	 // ifHasInstance()
+    	 Set<GKInstance> members = InstanceUtilities.getContainedInstances(instance,
```
Comment (guanmingwu) [inline src/org/gk/model/InstanceUtilities.java]: Please change members to another name, e.g. containedInstances

File: src/org/gk/model/InstanceUtilities.java [hunk 9]
```diff
@@ -1394,4 +1402,92 @@ else if (instance.getSchemClass().isa(ReactomeJavaConstants.Pathway))
...
+
+    	 // Check if instance contains a drug by:
+    	 //   (1) Iterating over all schema classes of the instance's members.
+    	 //   (2) Checking if any of the ancestor schema class is a Drug class.
+    	 return members.stream()
+					   .map(GKInstance::getSchemClass)
+					   .anyMatch(cls -> cls.isa(ReactomeJavaConstants.Drug));
+     }
+
+    private MySQLAdaptor getTestDBA() throws SQLException {
```
Comment (guanmingwu) [inline src/org/gk/model/InstanceUtilities.java]: Is this method called somewhere else? If not, don't add it.

I think it will be necessary for you to add another class for doing test. Otherwise, there is just too much code in the class.

File: src/org/gk/property/SearchDBTypeHelper.java [hunk 10]
```diff
@@ -101,35 +103,35 @@ public List mapGKInstanceToRenderable(String type,
                                         instance,
                                         id2Renderable);
         }
-        return new ArrayList(id2Renderable.values());
+        return new ArrayList<Renderable>(id2Renderable.values());
```
Comment (guanmingwu) [inline src/org/gk/property/SearchDBTypeHelper.java]: should: return new ArrayList<>(id2Renderable.values())?

File: src/org/gk/property/SearchDBTypeHelper.java [hunk 11]
```diff
@@ -101,35 +103,35 @@ public List mapGKInstanceToRenderable(String type,
...
-                                         Map id2Node) {
+                                         Map<Long, Renderable> id2Node) {
         try {
             // Stoichiometries are not supported as of Jan 11, 2007
-            List inputs = instance.getAttributeValuesList(ReactomeJavaConstants.input);
+            List<GKInstance> inputs = instance.getAttributeValuesList(ReactomeJavaConstants.input);
             if (inputs != null && inputs.size() > 0) {
-                for (Iterator it = inputs.iterator(); it.hasNext();) {
-                    GKInstance input = (GKInstance) it.next();
+                for (Iterator<GKInstance> it = inputs.iterator(); it.hasNext();) {
```
Comment (guanmingwu) [inline src/org/gk/property/SearchDBTypeHelper.java]: change to: for (GKInstance input : inputs) {

File: src/org/gk/render/AbstractNodeRenderer.java [hunk 12]
```diff
@@ -51,8 +54,13 @@ public void render(Graphics g) {
     }
 
     protected abstract void renderShapes(Graphics g);
-    
-    protected void renderDrug(Graphics g) {
+
+    protected void renderDrugLabel(Graphics g) {
+    	// Default to rectangle.
+		renderDrugLabel(g, DRUG_SHAPE.RECTANGLE);
```
Comment (guanmingwu) [inline src/org/gk/render/AbstractNodeRenderer.java]: As I said, please don't use DRUG_SHAPE. We may use different shapes for chemicals in the future. Therefore, this is not robust enough, and not flexible too.

File: src/org/gk/render/RenderableChemicalDrug.java [hunk 13]
```diff
@@ -1,16 +1,14 @@
...
 
-import java.awt.Color;
-
 /**
  * This customized Renderer is for ChemicalDrug.
  * @author wug
  *
  */
-public class RenderableChemicalDrug extends RenderableChemical {
+public class RenderableChemicalDrug extends RenderableChemical implements DefaultRenderConstants {
```
Comment (guanmingwu) [inline src/org/gk/render/RenderableChemicalDrug.java]: Don't add implements here. This has been taken care in supclass.

File: src/org/gk/render/RenderableComplex.java [hunk 14]
```diff
@@ -619,20 +619,7 @@ public boolean isAssignable(Renderable r) {
         }
         return false;
     }
-
```
Comment (guanmingwu) [inline src/org/gk/render/RenderableComplex.java]: Why did you delete this method?

File: src/org/gk/render/RenderableComplexDrug.java [hunk 15]
```diff
@@ -0,0 +1,20 @@
+package org.gk.render;
+
+public class RenderableComplexDrug extends RenderableComplex implements DefaultRenderConstants {
```
Comment (guanmingwu) [inline src/org/gk/render/RenderableComplexDrug.java]: No implements!

File: src/org/gk/render/RenderableComplexDrug.java [hunk 16]
```diff
@@ -0,0 +1,20 @@
...
+    /**
+     * Default constructor.
+     */
+    public RenderableComplexDrug() {
+        setForegroundColor(DEFAULT_DRUG_FORGROUND); // As the default for the drugs
```
Comment (guanmingwu) [inline src/org/gk/render/RenderableComplexDrug.java]: change to this(null)

File: src/org/gk/render/RenderableProteinDrug.java [hunk 17]
```diff
@@ -2,10 +2,10 @@
 
 import java.awt.Color;
 
-public class RenderableProteinDrug extends RenderableProtein {
+public class RenderableProteinDrug extends RenderableProtein implements DefaultRenderConstants {
```
Comment (guanmingwu) [inline src/org/gk/render/RenderableProteinDrug.java]: ??

File: src/org/gk/render/RenderableRNADrug.java [hunk 18]
```diff
@@ -2,10 +2,10 @@
 
 import java.awt.Color;
 
-public class RenderableRNADrug extends RenderableRNA {
+public class RenderableRNADrug extends RenderableRNA implements DefaultRenderConstants {
```
Comment (guanmingwu) [inline src/org/gk/render/RenderableRNADrug.java]: ??

File: src/org/gk/elv/InstanceZoomablePathwayEditor.java (hunk 2 above)
Comment (lbeckman314) [reply src/org/gk/elv/InstanceZoomablePathwayEditor.java]: Ah, sorry. In regards to the @Override and removed imports, I think that is my eclipse automatically formatting upon save. I will configure it so those don't occur. And I will change all tabs to 4 spaces!

File: src/org/gk/render/RenderableComplex.java (hunk 14 above)
Comment (lbeckman314) [reply src/org/gk/render/RenderableComplex.java]: Yes!

File: src/org/gk/render/RenderableComplexDrug.java (hunk 15 above)
Comment (lbeckman314) [reply src/org/gk/render/RenderableComplexDrug.java]: Understood!

File: src/org/gk/render/RenderableComplexDrug.java (hunk 16 above)
Comment (lbeckman314) [reply src/org/gk/render/RenderableComplexDrug.java]: Gotcha!

File: src/org/gk/elv/ElvReactionEditHelper.java (hunk 1 above)
Comment (lbeckman314) [reply src/org/gk/elv/ElvReactionEditHelper.java]: My mistake — will do!

File: src/org/gk/elv/InstanceZoomablePathwayEditor.java (hunk 3 above)
Comment (lbeckman314) [reply src/org/gk/elv/InstanceZoomablePathwayEditor.java]: Will do!

File: src/org/gk/elv/InstanceZoomablePathwayEditor.java (hunk 4 above)
Comment (lbeckman314) [reply src/org/gk/elv/InstanceZoomablePathwayEditor.java]: Ah, autoformatter mistake, sorry! I've changed my eclipse config to stop that behavior, should I reset the above change?

File: src/org/gk/property/SearchDBTypeHelper.java (hunk 11 above)
Comment (lbeckman314) [reply src/org/gk/property/SearchDBTypeHelper.java]: Will do.

File: src/org/gk/render/AbstractNodeRenderer.java (hunk 12 above)
Comment (lbeckman314) [reply src/org/gk/render/AbstractNodeRenderer.java]: Ah, my mistake. I missed that recommendation. I will change to something more robust.

File: src/org/gk/render/RenderableChemicalDrug.java (hunk 13 above)
Comment (lbeckman314) [reply src/org/gk/render/RenderableChemicalDrug.java]: Understood.

File: src/org/gk/render/RenderableRNADrug.java (hunk 18 above)
Comment (lbeckman314) [reply src/org/gk/render/RenderableRNADrug.java]: Fixed in [e77f0a9](https://github.com/reactome/CuratorTool/pull/23/commits/e77f0a99bd4c6699021ff050f23d7c451e9e7d35).

File: src/org/gk/render/RenderableProteinDrug.java (hunk 17 above)
Comment (lbeckman314) [reply src/org/gk/render/RenderableProteinDrug.java]: Fixed in [e77f0a9](https://github.com/reactome/CuratorTool/pull/23/commits/e77f0a99bd4c6699021ff050f23d7c451e9e7d35).

File: src/org/gk/render/RenderableComplexDrug.java (hunk 15 above)
Comment (lbeckman314) [reply src/org/gk/render/RenderableComplexDrug.java]: Fixed in [e77f0a9](https://github.com/reactome/CuratorTool/pull/23/commits/e77f0a99bd4c6699021ff050f23d7c451e9e7d35).

File: src/org/gk/render/RenderableChemicalDrug.java (hunk 13 above)
Comment (lbeckman314) [reply src/org/gk/render/RenderableChemicalDrug.java]: Fixed in [e77f0a9](https://github.com/reactome/CuratorTool/pull/23/commits/e77f0a99bd4c6699021ff050f23d7c451e9e7d35).

File: src/org/gk/property/SearchDBTypeHelper.java (hunk 11 above)
Comment (lbeckman314) [reply src/org/gk/property/SearchDBTypeHelper.java]: Fixed in [930d258](https://github.com/reactome/CuratorTool/pull/23/commits/930d25887be5b20f629ac6f51017f4806eec21a3).

File: src/org/gk/property/SearchDBTypeHelper.java (hunk 10 above)
Comment (lbeckman314) [reply src/org/gk/property/SearchDBTypeHelper.java]: Fixed in [b763f88](https://github.com/reactome/CuratorTool/pull/23/commits/b763f888eafed262c4bc2a5c257871707878643a).

File: src/org/gk/model/InstanceUtilities.java (hunk 9 above)
Comment (lbeckman314) [reply src/org/gk/model/InstanceUtilities.java]: Understood — fixed in [4082ae8](https://github.com/reactome/CuratorTool/pull/23/commits/4082ae86285bba3e8a3bf57c30f85e8735ec6773).

File: src/org/gk/model/InstanceUtilities.java (hunk 7 above)
Comment (lbeckman314) [reply src/org/gk/model/InstanceUtilities.java]: Done!

File: src/org/gk/model/InstanceUtilities.java (hunk 8 above)
Comment (lbeckman314) [reply src/org/gk/model/InstanceUtilities.java]: Fixed in [4779567](https://github.com/reactome/CuratorTool/pull/23/commits/477956733a5f9958693bca874094cb1914bcd684).

File: src/org/gk/gkCurator/authorTool/ComplexInstanceHandler.java (hunk 6 above)
Comment (lbeckman314) [reply src/org/gk/gkCurator/authorTool/ComplexInstanceHandler.java]: This is a good point! Fixed in (7f7d48d)[https://github.com/reactome/CuratorTool/pull/23/commits/7f7d48dae85ac51332a4bb83298257780b0a17de].

File: src/org/gk/elv/SimpleInstanceTransferHandler.java (hunk 5 above)
Comment (lbeckman314) [reply src/org/gk/elv/SimpleInstanceTransferHandler.java]: I think you're right! Removed autoformatter config from Eclipse.

File: src/org/gk/render/AbstractNodeRenderer.java (hunk 12 above)
Comment (lbeckman314) [reply src/org/gk/render/AbstractNodeRenderer.java]: Renamed to `RENDER_SHAPE` in [0ae7e0](https://github.com/reactome/CuratorTool/pull/23/commits/0ae7e0e6a2431ef5d9c1ca75f94794f93380743b).

File: src/org/gk/gkCurator/authorTool/InstanceHandlerNodeType.java [hunk 19]
```diff
@@ -0,0 +1,26 @@
+package org.gk.gkCurator.authorTool;
+
+import java.util.Map;
+
+import org.gk.model.GKInstance;
+import org.gk.property.SearchDBTypeHelper;
+import org.gk.render.Renderable;
+
+public class InstanceHandlerNodeType extends InstanceHandler {
```
Comment (guanmingwu) [inline src/org/gk/gkCurator/authorTool/InstanceHandlerNodeType.java]: Make it abstract

File: src/org/gk/model/InstanceUtilities.java [hunk 20]
```diff
@@ -1394,4 +1402,88 @@ else if (instance.getSchemClass().isa(ReactomeJavaConstants.Pathway))
...
+
+         List<String> validClasses = Arrays.asList(ReactomeJavaConstants.EntitySet,
+                                                   ReactomeJavaConstants.Complex);
+         // Check id instance is an EntitySet or Complex.
+         if (validClasses.stream().noneMatch(schemaClass::isa))
+             return false;
+
+         // If EntitySet has a "hasMember" or "hasCandidate" attribute then recursively iterate over them.
+         // ifHasInstance()
+         Set<GKInstance> containedInstances = InstanceUtilities.getContainedInstances(instance,
```
Comment (guanmingwu) [inline src/org/gk/model/InstanceUtilities.java]: no need to use InstanceUtiltlies here since you are in this class.

File: src/org/gk/property/SearchDBTypeHelper.java [hunk 21]
```diff
@@ -138,9 +139,9 @@ private void pulloutNodesForReaction(RenderableReaction reaction,
                 }
             }
             // Inhibitors
-            Collection regulations = InstanceUtilities.getRegulations(instance);
+            Collection<?> regulations = InstanceUtilities.getRegulations(instance);
```
Comment (guanmingwu) [inline src/org/gk/property/SearchDBTypeHelper.java]: Collection<GKInstance> regulations = InstanceUtilities.getRegulations(instance);

File: src/org/gk/render/Renderable.java [hunk 22]
```diff
@@ -26,7 +26,7 @@
  * pane.
  * @author  wgm
  */
-public abstract class Renderable implements Serializable, RenderablePropertyNames {
+public abstract class Renderable implements Serializable, RenderablePropertyNames, DefaultRenderConstants {
```
Comment (guanmingwu) [inline src/org/gk/render/Renderable.java]: Don't implement DefaultRenderConstants

**Comment by lbeckman314** [comment]
Superseded by #26.
